# LoG Book

## DiffAssemble-release Test Record

### 1. reference

- article link: https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Farxiv.org%2Fabs%2F2402.19302&data=05%7C02%7Cyanxia.lang%40durham.ac.uk%7C98a7a541f83e4d330a2308dd6d21290b%7C7250d88b4b684529be44d59a2d8a6f94%7C0%7C0%7C638786711364869299%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=6%2F4Zj%2FDKHg6OECDUN%2FSD2n8h7ckYbvsZHQjO11UFVBo%3D&reserved=0
- github link: https://github.com/IIT-PAVIS/DiffAssemble

### 2. DiffAssemble structure

- train_script 
  - .dataset 
    - dataset.dataset_utils
      - breakingbad_dt.py
      - celeba_dt.py
        - CelebA_HQ
      - nips_dt.py
        - Nips_dt
      - objects_dataset.py
        - Objects_Dataset
      - puzzle_dataset.py
        - Puzzle_Dataset
        - Puzzle_Dataset_MP
        - Puzzle_Dataset_Pad
        - Puzzle_Dataset_ROT
        - Puzzle_Dataset_ROT_MP
        - generate_random_expander
      - roc_dt.py
        - Roc_dt
      - sind_dt.py
        - Sind_dt
      - sind_vist_dt.py
        - Sind_Vist_dt
      - text_dataset.py
        - Text_dataset
      - vist_dataset.py
        - Vist_dataset
      - wiki_dt.py
        - Wiki_dt
      - wikiart_dt.py
        - Wikiart_DT
  - .model
    - spatial_diffusion_on_angle.py
      - .backbones
        - darknet_TransfConv.py
          - Dark_TFConv
        - __init__.py
        - efficient_gat.py
          - Eff_GAT
        - exophormer_gnn.py
          - Exophormer_GNN
  - ../.dataset
    - .data_splits
    - wikiart_subset_train.txt
            

### 3. git code install on NCC

#### using ncc terminal 
- platform: NCC COMP52715 Jupyter Server
- install code:
  - windows: open command -> Win+R -> input "cmd"
  - Mac/Linux: open terminal -> command + space -> input "Terminal"
  - folder upload: scp -r DiffAssemble-release remote-path

### 2. DiffAssemble trainning
#### i. preparation
- python version:
  - command:
    - python --version -> Python 2.7.18
    - python3 --version -> Python 3.8.10
    - reference: https://www.pythonpool.com/how-to-check-python-version/
  - notebook code:  

In [9]:
! python --version
! python3 --version
import sys 
print(sys.version)

import platform
print(platform.python_version())

Python 2.7.18
Python 3.8.10
3.11.11 (main, Dec  4 2024, 08:55:08) [GCC 9.4.0]
3.11.11


- python script run :
  - terminal: 
    - cd DiffAssemble-release/DiffAssemble-release/puzzle_diff
    - ~~ python train_script.py ~~
      - error: 
        - SyntaxError: invalid syntax -> 
          File "train_script.py", line 188
    experiment_name = f"{dataset}-{puzzle_sizes}-{steps}-degree:{degree}-virtnode:{virt_nodes}-{get_random_string(6)}-{'discrete' if discrete else 'continuous'}-{'random_dropout' if random_dropout else 'ours_dropout'}-arch-{architecture}"
        - fstring = f"" supported by python version > 3.6 
    - ~~ python3 train_script.py ~~
      - error: 
        - ModuleNotFoundError: No module named 'torch_geometric': File "train_script.py", line 5, in <module>
          import torch_geometric
        - pip install torch_geometric
        - ModuleNotFoundError: No module named 'torch'
        - pip install torch
        - ModuleNotFoundError: No module named 'pytorch_lightning'
        - pip install pytorch_lightning
        - AttributeError: module 'numpy' has no attribute 'typeDict'
          - introduced in numpy version 1.0.0, deprecated in numpy version 1.20.0, removed numpy in version 1.24.0
          - numpy.typedict didn’t provide efficient results with structured arrays, subarrays and other such numpy features.
          - numpy.sctypeDict has been introduced as an alternative to the typedict attribute.
          - reference: https://www.pythonpool.com/fixed-module-numpy-has-no-attribute-typedict/
          - check numpy version: 
            - only show numpy: pip show numpy -> Name: numpy Version: 1.24.4
            - show all python packages: pip list
            - using python code


In [1]:
import numpy
print(numpy.__version__)

1.26.3


        - scipy old version caused:
          - upgrade version: python3 -m pip install --upgrade scipy => version 1.3.3 -> 1.10.1
        - error caused by import dataset_utils 
          - ModuleNotFoundError: No module named 'torchvision'
          - pip install torchvision
          - error caused by import breakingbad_dt
            - ModuleNotFoundError: No module named 'trimesh'
            - pip install trimesh => 4.6.13
          - error caused by import objects_dataset
            - ModuleNotFoundError: No module named 'einops'
            - pip install einops => 0.8.1
            - ImportError: cannot import name 'Resampling' from 'PIL.Image'
              - supported by pillow version > 9.0.0
              - upgrade version: python3 -m pip install --upgrade pillow => version 7.0.0 -> 10.4.0
          - error caused by import wiki_dt

In [7]:
from PIL import Image
from PIL.Image import Resampling

print(Image.__version__)
print(Resampling.__version__)

10.2.0


AttributeError: __version__

          - error caused by import wiki_dt
            - ModuleNotFoundError: No module named 'pandas'
            - pip install pandas => 2.0.3
            - ModuleNotFoundError: No module named 'sklearn'
            - pip install sklearn
              - The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
              - pip install scikit-learn => 1.3.2
        - TypeError: 'type' object is not subscriptable
          - supported by python version >= 3.9
          - python3.9 train_script.py - python3.12 train_script.py
            - ModuleNotFoundError: No module named 'torch_geometric'
            - pip -V
              - pip 20.0.2 from /usr/lib/python3/dist-packages/pip (python 3.8)
              - default link to python3.8
            - python3.9 -m pip install torch_geometric
            - !!! different python versions cannot share the modules 

- the usage of the virtual environment 
  - benefits: 
    - ~~shared modules~~
      - only pure python modulels available to share with different python version
        - source code open and read directly 
        - file extension: only contain .py file 
      - python modules compiled for a specific version
        - file extension: .dll .so .pyd etc
        - e.g.1 NumPy - C code, 
        - e.g.2 panda - depend on NumPy and Cython
        - e.g.3 SciPy - Built on top of NumPy with C/C++ and Fortran code
        - e.g.4 scikit-learn - Uses Cython and compiled C/C++ for performance.
        - e.g.5 TensorFlow - Core engine is in C++ with Python bindings
        - e.g.6 PyTorch - Backend is C++ (ATen) with Python interface.
    - environmants copied
    - differnet projects supported by different version of modules or python
  - create virtual environment
    - python3 -m venv venv
    - ~~python -m venv venv~~
      - unavailable for python version 2
  - enter virtual environment
    - linux or macOS: source venv/bin/activate   
    - windows: venv\Scripts\activate  
  - copy environment
    - pip freeze > requirements.txt
    - pip install -r requirements.txt
  - exit environment
    - deactivate
  - delete environment
    - linux or macOS: rm -rf env_name 
    - windows: rmdir /s /q env_name 
- module download and install 
  - .whl file: 
    - binary package
    - ZIP archive including:
      - Python code (.py)
      - Compiled extensions (.pyd, .so)
      - Metadata (METADATA, RECORD, etc.)
    - e.g. numpy‑1.25.0‑cp311‑cp311‑win_amd64.whl
      - numpy version 1.25.0
      - Built for CPython 3.11
      - Windows 64-bit
    - advantages: 
      - Faster installation:
        - fast to install without compile from source 
      - Platform-specific: 
        - tailored for Windows, macOS, Linux, and specific Python versions
      - Offline installation: 
        - download and install without internet access.
  - install modules depended

   - python3.12 virtual environment test
     - python3.12 -m venv venv312
     - source venv312/bin/activate
     - python --version -> 3.12.9
     - ~~python train_script.py~~
       - error:
        - ModuleNotFoundError: No module named 'torch_geometric'
        - pip install torch_geometric -> 2.6.1
        - ModuleNotFoundError: No module named 'torch'
        - pip install torch -> 2.7.1
        - ModuleNotFoundError: No module named 'pytorch_lightning'
        - pip install pytorch_lightning -> 2.5.2
        - ModuleNotFoundError: No module named 'torchvision'
        - pip install torchvision -> 0.22.1
        - ModuleNotFoundError: No module named 'trimesh'
        - pip install trimesh -> 4.6.13
        - ModuleNotFoundError: No module named 'scipy'
        - pip install scipy -> 1.16.0
        - ModuleNotFoundError: No module named 'einops'
        - pip install einops -> 0.8.1
        - ModuleNotFoundError: No module named 'matplotlib'
        - pip install matplotlib -> 3.10.3
        - ModuleNotFoundError: No module named 'pandas'
        - pip install pandas -> 2.3.0
        - ModuleNotFoundError: No module named 'sklearn'
        - pip install scikit-learn -> 1.7.0
        - ModuleNotFoundError: No module named 'timm'
        - pip install timm -> 1.0.16
        - ModuleNotFoundError: No module named 'kornia'
        - pip install kornia ->  0.8.1
        - ModuleNotFoundError: No module named 'transformers'
        - pip install transformers -> 4.53.1
        - ModuleNotFoundError: No module named 'wandb'
        - pip install wandb -> 0.21.0
        - ModuleNotFoundError: No module named 'model.backbones.backbone_vist'
          - no backbone_vist.py find -> comment code
        - ModuleNotFoundError: No module named 'torch_scatter'
        - pip install torch_scatter
          -   Using cached torch_scatter-2.1.2.tar.gz (108 kB)
              Installing build dependencies ... done
              Getting requirements to build wheel ... error
              - File "/tmp/pip-build-env-ud0ncm91/overlay/lib/python3.12/site-packages/setuptools/build_meta.py", line 317, in run_setup
          exec(code, locals())
        File "<string>", line 8, in <module>
      ModuleNotFoundError: No module named 'torch'
          - check torch version
            - python
            - import torch
            - print(torch.__version__)
            - 2.7.1 + cu126
            - exit -> contrl+D
          - pip install torch-scatter -f https://data.pyg.org/whl/torch-2.7.1+cu126.html
            - the lastest verison supports 2.7.0, 2.7.1 not find
              - build torch-scatter from source
              - pip install setuptools wheel ninja cmake -> wheel-0.45.1,ninja-1.11.1.4,cmake-4.0.3
              - git clone https://github.com/rusty1s/pytorch_scatter.git
              - cd pytorch_scatter   
              - pip install . -> 2.1.2
             
        - pip install torch_scatter
        - ModuleNotFoundError: No module named 'pytorch3d'
        - pip install pytorch3d
          - the lastest version supports pytorch 2.7.1, no 2.7.1
          - rebuild from source becuase of no prebuilt wheels
            - git clone https://github.com/facebookresearch/pytorch3d.git
            - cd pytorch3d
            - pip install . -> 0.7.8
        - FileNotFoundError: [Errno 2] No such file or directory: 'datasets/data_splits/wikiart_subset_train.txt'
          - puzzle_diff/train_script.py", line 114, in main
    train_dt, test_dt, puzzle_sizes = du.get_dataset
          - puzzle_diff/dataset/dataset_utils.py", line 74, in get_dataset
    train_dt = Wikiart_DT(train=True)
          - puzzle_diff/dataset/wikiart_dt.py - line 18, in __init__
    with open(train_file, "r") as f:
          - cd ../

   - python puzzle_diff/train_script.py

  - jupyter notebook: path-same root with puzzle_diff
    - code:  

In [ ]:
%run puzzle_diff/train_script.py 

     - error: 
       - No module named 'torch_geometric'
     - code: pip install torch_geometric
     - error: 
       - Permission denied

In [2]:
import pytorch_lightning as pl

Python 3.8.10


aiohappyeyeballs         2.4.4
aiohttp                  3.10.11
aiosignal                1.3.1
appdirs                  1.4.3
async-timeout            5.0.1
attrs                    19.3.0
Automat                  0.8.0
blessings                1.7
blinker                  1.4
certifi                  2019.11.28
chardet                  3.0.4
Click                    7.0
cloud-init               24.4.1
ClusterShell             1.8.3
colorama                 0.4.3
command-not-found        0.3
configobj                5.0.6
constantly               15.1.0
cryptography             2.8
cupshelpers              1.0
cycler                   0.10.0
dbus-python              1.2.16
decorator                4.4.2
defer                    1.0.6
distlib                  0.3.0
distro                   1.4.0
distro-info              0.23+ubuntu1.1
einops                   0.8.1
entrypoints              0.3
filelock                 3.0.12
frozenlist               1.5.0
fsspec                   2025.3.0
GDAL                     3.0.4
gpustat                  0.4.1
greenlet                 0.4.15
httplib2                 0.14.0
hyperlink                19.0.0
idna                     2.8
importlib-metadata       1.5.0
incremental              16.10.1
iotop                    0.6
Jinja2                   2.10.1
jsonpatch                1.22
jsonpointer              2.0
jsonschema               3.2.0
keyring                  18.0.1
kiwisolver               1.0.1
language-selector        0.1
launchpadlib             1.10.13
lazr.restfulclient       0.14.2
lazr.uri                 1.0.3
lightning-utilities      0.11.9
macaroonbakery           1.3.1
Mako                     1.1.0
Markdown                 3.1.1
MarkupSafe               1.1.0
matplotlib               3.1.2
more-itertools           4.2.0
mpmath                   1.3.0
msgpack                  0.6.2
multidict                6.1.0
netifaces                0.10.4
networkx                 3.1
numpy                    1.24.4
nvidia-cublas-cu12       12.1.3.1
nvidia-cuda-cupti-cu12   12.1.105
nvidia-cuda-nvrtc-cu12   12.1.105
nvidia-cuda-runtime-cu12 12.1.105
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.0.2.54
nvidia-curand-cu12       10.3.2.106
nvidia-cusolver-cu12     11.4.5.107
nvidia-cusparse-cu12     12.1.0.106
nvidia-ml-py             7.352.0
nvidia-ml-py3            7.352.0
nvidia-nccl-cu12         2.20.5
nvidia-nvjitlink-cu12    12.9.86
nvidia-nvtx-cu12         12.1.105
oauthlib                 3.1.0
olefile                  0.46
packaging                20.3
pexpect                  4.6.0
Pillow                   7.0.0
pip                      20.0.2
propcache                0.2.0
protobuf                 3.6.1
psutil                   5.9.0
pyasn1                   0.4.2
pyasn1-modules           0.2.1
pycairo                  1.16.2
pycups                   1.9.73
Pygments                 2.3.1
PyGObject                3.36.0
PyHamcrest               1.9.0
PyJWT                    1.7.1
pymacaroons              0.13.0
PyNaCl                   1.3.0
pynvim                   0.4.1
pyOpenSSL                19.0.0
pyparsing                2.4.6
pyRFC3339                1.1
pyrsistent               0.15.5
pyserial                 3.4
python-apt               2.0.1+ubuntu0.20.4.1
python-dateutil          2.7.3
python-debian            0.1.36+ubuntu1.1
pytorch-lightning        2.4.0
pytz                     2019.3
PyYAML                   6.0.2
requests                 2.22.0
requests-unixsocket      0.2.0
scipy                    1.10.1
screen-resolution-extra  0.0.0
SecretStorage            2.3.1
service-identity         18.1.0
setuptools               45.2.0
simplejson               3.16.0
six                      1.14.0
sos                      4.7.2
ssh-import-id            5.10
sympy                    1.13.3
systemd-python           234
torch                    2.4.1
torch-geometric          2.6.1
torchmetrics             1.5.2
torchvision              0.19.1
tqdm                     4.67.1
trimesh                  4.6.13
triton                   3.0.0
Twisted                  18.9.0
typing-extensions        4.13.2
ubuntu-pro-client        8001
ufw                      0.36
unattended-upgrades      0.1
urllib3                  1.25.8
virtualenv               20.0.17
wadllib                  1.3.3
wheel                    0.34.2
xkit                     0.0.0
yarl                     1.15.2
zipp                     1.0.0
zope.interface           4.7.1